source/components/Autocomplete.js

```diff
@@ -0,0 +1,190 @@
+
+  closeSuggestions = () => {
+    this.setState({ isSuggestionsOpened: false, highlightedOptionIndex: 0 });
+  };
+
+  handleAutocompleteClick = () => {
+    this._getInputSkinPart().focus();
+    if (!this.state.isSuggestionsOpened) {
+      this.openSuggestions();
+    } else {
+      this.closeSuggestions();
+    }
+  };
+
+  _handleDocumentClick = (event) => {
+    const root = this._getRootSkinPart();
+    const isDescendant = events.targetIsDescendant(event, ReactDOM.findDOMNode(root));
+    if (this.state.isSuggestionsOpened && !isDescendant) {
+      this.setState({ isSuggestionsOpened: false, highlightedOptionIndex: 0 });
+    }
+  };
+
+  onKeyDown = (event) => {
+    const { selectedWords } = this.state;
+    let keyPressed = event.which;
+
+    switch (event.keyCode) {
+      case 8: // Delte on backspace
+        if (!event.target.value && selectedWords.length) {
+          this.removeWord(selectedWords[selectedWords.length - 1], event);
+        }
+        break;
+      case 9: // Select currently highlighted option on 'Tab' key
+        event.preventDefault();
+        this.updateSelectedWords(event);
+        break;
+      case 13: // Select currently highlighted option on 'Enter' key
+        this.updateSelectedWords(event);
+        break;
+      case 27: // Close on 'Escape' key
+        this.closeSuggestions();
+        break;
+      case 38: // Move selection higlight 'up' on Arrow Up key
+        this._handleHighlightMove('up');
+        break;
+      case 40: // Move selection higlight 'down' on Arrow Down key
+        this._handleHighlightMove('down');
+        break;
+      default:
+        this.openSuggestions();
+    }
+  };
+
+  _handleHighlightMove = (direction) => {
+    let position;
+    if (direction === 'up') {
+      position = this.state.highlightedOptionIndex - 1;
+    } else if (direction === 'down') {
+      position = this.state.highlightedOptionIndex + 1;
+    }
+    if (position >= 0 && position < this.state.filteredWords.length) {
+      this.setState({ highlightedOptionIndex: position, isSuggestionsOpened: true });
+    }
+  };
+
+  setHighlightedOptionIndex = (index) => {
+    this.setState({ highlightedOptionIndex: index });
+  };
+
+  // onKeyUp handler
+  clearInvalidChars = (event) => {
+    let value = event.target.value;
+
+    if (this.state.invalidCharacters.test(value)) {
+      event.target.value = value.replace(this.state.invalidCharacters, '');
+      return;
+    }
+
+    const filteredWords = [];
+    _.some(this.props.suggestedWords, function (suggestedWord) {
+      if (_.startsWith(suggestedWord, value)) {
+        filteredWords.push(suggestedWord);
+      }
+    });
+
+    this.setState({ filteredWords: (value !== '') ? filteredWords : this.props.suggestedWords });
+  };
+
+  updateSelectedWords = (event) => {
+    if (!this.props.maxSelections || (this.state.selectedWords.length < this.props.maxSelections && this.state.filteredWords && this.state.filteredWords.length > 0)) {
```

Comment:
Please shorten this condition with separate boolean vars like

`const canMoreWordsBeSelected = this.state.selectedWords.length < this.props.maxSelections`